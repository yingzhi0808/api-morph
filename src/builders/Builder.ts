/**
 * 构建器基础接口，定义所有构建器的通用契约
 *
 * @category Builders
 */
export interface Builder<T> {
  /**
   * 构建最终对象
   * @returns 构建完成的对象
   */
  build(): T;
}
