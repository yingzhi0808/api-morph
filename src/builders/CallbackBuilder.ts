import { cloneDeep } from "radashi";
import type { Builder } from "@/core";
import type { CallbackObject, PathItemObject } from "@/types";

/**
 * 回调构建器，用于构建 OpenAPI CallbackObject
 */
export class CallbackBuilder implements Builder<CallbackObject> {
  private callback: CallbackObject = {};

  build() {
    return cloneDeep(this.callback);
  }

  /**
   * 添加回调表达式。
   * @param expression 运行时计算的表达式，用于标识回调操作的URL。
   * @param pathItem 描述一组请求和预期响应的路径项对象。
   * @returns 回调构建器。
   */
  addExpression(expression: string, pathItem: PathItemObject) {
    const callback = this.callback;
    if (!callback[expression]) callback[expression] = pathItem;
    return this;
  }

  /**
   * 添加扩展字段。
   * @param key 扩展字段键（必须以 'x-' 开头）。
   * @param value 扩展字段值。
   * @returns 回调构建器。
   */
  addExtension(key: `x-${string}`, value: unknown) {
    const document = this.callback;
    if (!document[key]) document[key] = value;
    return this;
  }
}
