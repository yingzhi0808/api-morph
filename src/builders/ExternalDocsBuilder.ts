import { cloneDeep } from "radashi";
import type { Builder } from "@/core";
import type { ExternalDocumentationObject } from "@/types";

/**
 * 外部文档构建器，用于构建 OpenAPI ExternalDocumentationObject
 */
export class ExternalDocsBuilder implements Builder<ExternalDocumentationObject> {
  private externalDocs: ExternalDocumentationObject = { url: "" };

  build() {
    return cloneDeep(this.externalDocs);
  }

  /**
   * 设置外部文档URL。
   * @param url 外部文档URL（必需）。
   * @returns 外部文档构建器。
   */
  setUrl(url: string) {
    this.externalDocs.url = url;
    return this;
  }

  /**
   * 设置外部文档描述。
   * @param description 外部文档描述。
   * @returns 外部文档构建器。
   */
  setDescription(description: string) {
    this.externalDocs.description = description;
    return this;
  }

  /**
   * 添加扩展字段。
   * @param key 扩展字段键（必须以 'x-' 开头）。
   * @param value 扩展字段值。
   * @returns 响应构建器。
   */
  addExtension(key: `x-${string}`, value: unknown) {
    const externalDocs = this.externalDocs;
    if (!externalDocs[key]) externalDocs[key] = value;
    return this;
  }
}
