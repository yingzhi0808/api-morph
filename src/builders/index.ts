/* v8 ignore start */

export * from "./Builder";
export * from "./CallbackBuilder";
export * from "./DocumentBuilder";
export * from "./ExternalDocsBuilder";
export * from "./OperationBuilder";
export * from "./ParameterBuilder";
export * from "./PathItemBuilder";
export * from "./RequestBodyBuilder";
export * from "./ResponseBuilder";
export * from "./SecurityBuilder";
export * from "./ServerBuilder";
