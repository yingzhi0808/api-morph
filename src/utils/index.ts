/* v8 ignore start */

export * from "./string";
export * from "./typeGuards";
