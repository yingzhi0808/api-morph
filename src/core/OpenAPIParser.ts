import type { Project, SourceFile } from "ts-morph";
import { SyntaxKind } from "typescript";
import { ExpressRouteAnalyzer } from "@/analyzers";
import { type OpenAPIBuilder, PathItemBuilder } from "@/builders";
import { JSDocTagName } from "@/constants";
import {
  CallbackTagParser,
  DeprecatedTagParser,
  DescriptionTagParser,
  ExtensionsTagParser,
  ExternalDocsTagParser,
  OperationIdTagParser,
  OperationTagParser,
  ParameterTagParser,
  RequestBodyTagParser,
  ResponsesExtensionsTagParser,
  ResponseTagParser,
  SecurityTagParser,
  ServerTagParser,
  SimplifiedResponseTagParser,
  SummaryTagParser,
  TagsTagParser,
} from "@/parsers";
import type { ParseContext, ParsedOperation, ParserOptions, SourceOperationData } from "@/types";
import { ASTAnalyzerRegistry } from "./ASTAnalyzerRegistry";
import { OperationComposer } from "./OperationComposer";
import { TagParserRegistry } from "./TagParserRegistry";

/**
 * 解析整个项目中的 JSDoc 注释，并生成 OpenAPI 对象
 */
export class OpenAPIParser {
  /** 解析上下文 */
  private context: ParseContext;
  /** 标签解析器注册表 */
  private tagParserRegistry: TagParserRegistry;
  /** AST分析器注册表 */
  private astAnalyzerRegistry: ASTAnalyzerRegistry;

  constructor(
    private project: Project,
    private options: ParserOptions = {},
  ) {
    this.context = this.createParseContext(this.project, options);
    this.tagParserRegistry = this.createTagParserRegistry(this.options);
    this.astAnalyzerRegistry = this.createASTAnalyzerRegistry(this.options);
  }

  /**
   * 解析整个项目，返回完整的 OpenAPI 对象。
   * @param openAPIBuilder 文档构建器实例。
   * @returns 返回完整的 OpenAPI 对象。
   */
  async parse(openAPIBuilder: OpenAPIBuilder) {
    const operationParser = new OperationComposer(this.tagParserRegistry, this.astAnalyzerRegistry);

    const include = this.context.options.include;
    const excludes = this.context.options.exclude;

    let sourceFiles: SourceFile[];
    if (Array.isArray(include)) {
      if (include.length === 0) {
        sourceFiles = this.project.getSourceFiles([]);
      } else {
        const patterns = [...include];
        if (Array.isArray(excludes) && excludes.length > 0) {
          patterns.push(...excludes.map((pattern) => `!${pattern}`));
        }
        sourceFiles = this.project.getSourceFiles(patterns);
      }
    } else if (Array.isArray(excludes) && excludes.length > 0) {
      const patterns: string[] = excludes.map((pattern) => `!${pattern}`);
      sourceFiles = this.project.getSourceFiles(patterns);
    } else {
      sourceFiles = this.project.getSourceFiles();
    }

    const flatOperationDataList = sourceFiles.flatMap((sourceFile) =>
      this.findSourceOperationDataList(sourceFile),
    );

    const results = await Promise.all(
      flatOperationDataList.map((sourceOperationData) =>
        operationParser.compose(sourceOperationData),
      ),
    );

    const filteredResults = this.filterDeprecatedOperations(results);
    const openAPIObject = this.buildOpenAPI(filteredResults, openAPIBuilder);
    return openAPIObject;
  }

  /**
   * 根据 includeDeprecated 选项过滤废弃的操作。
   * @param parsedOperationList 解析得到的操作数组。
   * @returns 过滤后的操作数组。
   */
  private filterDeprecatedOperations(parsedOperationList: ParsedOperation[]): ParsedOperation[] {
    if (this.context.options.includeDeprecated) {
      return parsedOperationList;
    }
    return parsedOperationList.filter(
      (parsedOperationData) => !parsedOperationData.operation.deprecated,
    );
  }

  /**
   * 构建完整的 OpenAPI 文档。
   * @param parsedOperationList 解析得到的操作数组。
   * @param openAPIBuilder 预构建文档构建器实例。
   * @returns OpenAPI 文档对象。
   */
  private buildOpenAPI(parsedOperationList: ParsedOperation[], openAPIBuilder: OpenAPIBuilder) {
    // 直接使用传入的 OpenAPIBuilder 实例
    const builder = openAPIBuilder;

    const tags = new Set<string>();
    const pathItemBuilders = new Map<string, PathItemBuilder>();

    for (const parsedOperationData of parsedOperationList) {
      const { path, method, operation } = parsedOperationData;

      let pathItemBuilder = pathItemBuilders.get(path);
      if (!pathItemBuilder) {
        pathItemBuilder = new PathItemBuilder();
        pathItemBuilders.set(path, pathItemBuilder);
      }

      pathItemBuilder.addOperation(method, operation);

      if (operation.tags) {
        operation.tags.forEach((tag) => tags.add(tag));
      }
    }

    for (const [path, pathItemBuilder] of pathItemBuilders.entries()) {
      builder.addPathItemFromBuilder(path, pathItemBuilder);
    }

    tags.forEach((tag) => {
      builder.addTag({ name: tag });
    });

    for (const [name, schema] of this.context.schemas) {
      builder.addSchemaToComponents(name, schema);
    }

    return builder.build();
  }

  /**
   * 创建标签解析器注册表。
   * @param options 配置选项。
   * @returns 标签解析器注册表。
   */
  private createTagParserRegistry(options: ParserOptions) {
    const tagParserRegistry = new TagParserRegistry();
    const defaultParsers = [
      SecurityTagParser,
      ServerTagParser,
      ParameterTagParser,
      RequestBodyTagParser,
      ResponseTagParser,
      CallbackTagParser,
      ExternalDocsTagParser,
      OperationIdTagParser,
      DeprecatedTagParser,
      ExtensionsTagParser,
      ResponsesExtensionsTagParser,
      DescriptionTagParser,
      SummaryTagParser,
      TagsTagParser,
      OperationTagParser,
      SimplifiedResponseTagParser,
    ];
    const parsers = [...defaultParsers, ...(options.customParsers ?? [])];

    for (const parser of parsers) {
      tagParserRegistry.register(new parser(this.context));
    }

    return tagParserRegistry;
  }

  /**
   * 创建AST分析器注册表。
   * @param options 配置选项。
   * @returns AST分析器注册表。
   */
  private createASTAnalyzerRegistry(options: ParserOptions) {
    const astAnalyzerRegistry = new ASTAnalyzerRegistry();

    // 如果禁用了AST分析，返回空的注册表
    if (options.enableASTAnalysis === false) {
      return astAnalyzerRegistry;
    }

    const defaultAnalyzers = [ExpressRouteAnalyzer];
    const analyzers = [...defaultAnalyzers, ...(options.customAnalyzers ?? [])];

    for (const analyzer of analyzers) {
      astAnalyzerRegistry.register(new analyzer(this.context));
    }

    return astAnalyzerRegistry;
  }

  /**
   * 创建解析上下文。
   * @param project ts-morph Project 实例。
   * @param parserOptions 自定义解析选项。
   * @returns 解析上下文实例。
   */
  private createParseContext(project: Project, parserOptions: ParserOptions): ParseContext {
    const typeChecker = project.getTypeChecker();

    // 默认选项
    const defaultOptions: ParserOptions = {
      includeDeprecated: true,
      defaultResponseMediaType: "application/json",
      defaultRequestMediaType: "application/json",
      enableASTAnalysis: true,
    };

    // 合并用户选项和默认选项
    const options: ParserOptions = { ...defaultOptions, ...parserOptions };

    return {
      project,
      typeChecker,
      schemas: new Map(),
      options,
    };
  }

  /**
   * 查找指定 TypeScript 文件中所有被 @operation 标签注释的 AST 节点。
   * @param sourceFile 要解析的 TypeScript 源文件。
   * @returns 返回一个 SourceOperationData 数组，其中每个对象包含被 JSDoc 标签注释的 AST 节点及其所有标签，
   * 如果未找到匹配的节点，则返回空数组。
   */
  private findSourceOperationDataList(sourceFile: SourceFile) {
    const jsDocNodes = sourceFile.getDescendantsOfKind(SyntaxKind.JSDoc);
    const sourceOperationDataList: SourceOperationData[] = [];

    for (const jsDocNode of jsDocNodes) {
      const tags = jsDocNode.getTags();

      // 检查是否包含 @operation 标签
      const hasOperation = tags.some((tag) => tag.getTagName() === JSDocTagName.OPERATION);
      if (!hasOperation) {
        continue;
      }

      // 检查是否包含 @hidden 标签，如果有则跳过此操作
      const isHidden = tags.some((tag) => tag.getTagName() === JSDocTagName.HIDDEN);
      if (isHidden) {
        continue;
      }

      const parentNode = jsDocNode.getParent();
      if (parentNode) {
        sourceOperationDataList.push({
          node: parentNode,
          tags,
        });
      }
    }

    return sourceOperationDataList;
  }
}
