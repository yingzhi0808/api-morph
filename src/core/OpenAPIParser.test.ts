import { createFileWithContent, createProject } from "@tests/utils";
import { describe, expect, it } from "vitest";
import { OpenAPIBuilder } from "@/builders";
import { OpenAPIParser } from "./OpenAPIParser";

describe("OpenAPIParser", () => {
  describe("标签解析功能测试", () => {
    it("应该解析包含JSON Schema的JSDoc注释", async () => {
      const project = createProject();
      createFileWithContent(
        project,
        "user-controller.ts",
        `
/**
 * @operation put /api/v1/users/{userId}/profile
 * @summary 更新用户档案信息
 * @description 更新指定用户的个人档案信息，包括基本信息和偏好设置。支持部分更新，只更新提供的字段。
 * @operationId updateUserProfile
 * @tags users profile
 * @deprecated
 * @parameter userId path required 用户唯一标识符
 * schema:
 *   type: string
 *   format: uuid
 * example: 123e4567-e89b-12d3-a456-426614174000
 * @requestBody required 用户档案更新数据
 * content:
 *   application/json:
 *     schema:
 *       type: object
 *       required: [email]
 *       properties:
 *         email:
 *           type: string
 *           format: email
 *           description: 用户邮箱地址
 *           examples: [john.doe@example.com]
 *         displayName:
 *           type: string
 *           minLength: 2
 *           maxLength: 50
 *           description: 用户显示名称
 *           examples: [John Doe]
 *         avatar:
 *           type: string
 *           format: uri
 *           description: 头像URL
 *           examples: [https://cdn.example.com/avatars/user123.jpg]
 * @response 200 用户档案更新成功
 * content:
 *   application/json:
 *     schema:
 *       type: object
 *       properties:
 *         success:
 *           type: boolean
 *           description: 是否成功
 *           examples: [true]
 *         data:
 *           type: object
 *           description: 返回数据
 *           properties:
 *             userId:
 *               type: string
 *               format: uuid
 *               description: 用户ID
 *               examples: [123e4567-e89b-12d3-a456-426614174000]
 * @response 404 用户不存在
 * content:
 *   application/json:
 *     schema:
 *       type: object
 *       properties:
 *         success:
 *           type: boolean
 *           description: 是否成功
 *           examples: [false]
 *         error:
 *           type: object
 *           properties:
 *             code:
 *               type: string
 *               description: 错误码
 *               examples: [USER_NOT_FOUND]
 *             message:
 *               type: string
 *               description: 错误消息
 *               examples: [指定的用户不存在]
 */
app.put("/api/v1/users/:userId/profile", (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder();
      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });

    it("应该解析包含内联schema的JSDoc注释", async () => {
      const project = createProject();
      createFileWithContent(
        project,
        "user-controller.ts",
        `
/**
 * @operation put /api/v1/users/{userId}/profile
 * @summary 更新用户档案信息
 * @description 更新指定用户的个人档案信息，包括基本信息和偏好设置。支持部分更新，只更新提供的字段。
 * @operationId updateUserProfile
 * @tags users profile
 * @deprecated
 * @parameter userId path required 用户唯一标识符
 * schema:
 *   type: string
 *   format: uuid
 * example: 123e4567-e89b-12d3-a456-426614174000
 * @requestBody {$ref: "#/components/schemas/UpdateUserProfileDto"} required 用户档案更新数据
 * @response 200 {$ref: "#/components/schemas/UpdateUserProfileVo"} 用户档案更新成功
 * @response 404 {$ref: "#/components/schemas/UserNotFoundVo"} 用户不存在
 */
app.put("/api/v1/users/:userId/profile", (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder()
        .addSchemaToComponents("UpdateUserProfileDto", {
          type: "object",
          required: ["email"],
          properties: {
            email: {
              type: "string",
              format: "email",
              description: "用户邮箱地址",
              examples: ["john.doe@example.com"],
            },
            displayName: {
              type: "string",
              minLength: 2,
              maxLength: 50,
              description: "用户显示名称",
              examples: ["John Doe"],
            },
            avatar: {
              type: "string",
              format: "uri",
              description: "头像URL",
              examples: ["https://cdn.example.com/avatars/user123.jpg"],
            },
          },
        })
        .addSchemaToComponents("UpdateUserProfileVo", {
          type: "object",
          properties: {
            success: {
              type: "boolean",
              description: "是否成功",
              examples: [true],
            },
            data: {
              type: "object",
              description: "返回数据",
              properties: {
                userId: {
                  type: "string",
                  format: "uuid",
                  description: "用户ID",
                  examples: ["123e4567-e89b-12d3-a456-426614174000"],
                },
              },
            },
          },
        })
        .addSchemaToComponents("UserNotFoundVo", {
          type: "object",
          properties: {
            success: {
              type: "boolean",
              description: "是否成功",
              examples: [false],
            },
            error: {
              type: "object",
              properties: {
                code: {
                  type: "string",
                  description: "错误码",
                  examples: ["USER_NOT_FOUND"],
                },
                message: {
                  type: "string",
                  description: "错误消息",
                  examples: ["指定的用户不存在"],
                },
              },
            },
          },
        });

      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });

    it("应该解析包含@link引用的Zod schema的JSDoc注释", async () => {
      const project = createProject({
        tsConfigFilePath: "tsconfig.json",
        useInMemoryFileSystem: false,
        skipAddingFilesFromTsConfig: true,
      });
      project.addDirectoryAtPath("tests/fixtures");
      createFileWithContent(
        project,
        "user-controller.ts",
        `
import { UpdateUserProfileDto, UpdateUserProfileVo, UserNotFoundVo } from "@tests/fixtures/schema"
/**
 * @operation put /api/v1/users/{userId}/profile
 * @summary 更新用户档案信息
 * @description 更新指定用户的个人档案信息，包括基本信息和偏好设置。支持部分更新，只更新提供的字段。
 * @operationId updateUserProfile
 * @tags users profile
 * @deprecated
 * @parameter userId path required 用户唯一标识符
 * schema:
 *   type: string
 *   format: uuid
 * example: 123e4567-e89b-12d3-a456-426614174000
 * @requestBody {@link UpdateUserProfileDto} required 用户档案更新数据
 * @response 200 {@link UpdateUserProfileVo} 用户档案更新成功
 * @response 404 {@link UserNotFoundVo} 用户不存在
 */
app.put("/api/v1/users/:userId/profile", (req, res) => {})`,
      );

      const parser = new OpenAPIParser(project);
      const builder = new OpenAPIBuilder();
      const result = await parser.parse(builder);

      expect(result).toMatchSnapshot();
    });
  });
});
