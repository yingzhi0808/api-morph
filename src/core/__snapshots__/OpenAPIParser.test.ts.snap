// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OpenAPIParser > 标签解析功能测试 > 应该解析包含@link引用的Zod schema的JSDoc注释 1`] = `
{
  "components": {
    "schemas": {
      "UpdateUserProfileDto": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "avatar": {
            "description": "头像URL",
            "examples": [
              "https://cdn.example.com/avatars/user123.jpg",
            ],
            "format": "uri",
            "type": "string",
          },
          "displayName": {
            "description": "用户显示名称",
            "examples": [
              "John Doe",
            ],
            "maxLength": 50,
            "minLength": 2,
            "type": "string",
          },
          "email": {
            "description": "用户邮箱地址",
            "examples": [
              "john.doe@example.com",
            ],
            "format": "email",
            "pattern": "^(?!\\.)(?!.*\\.\\.)([A-Za-z0-9_'+\\-\\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\\-]*\\.)+[A-Za-z]{2,}$",
            "type": "string",
          },
        },
        "required": [
          "email",
        ],
        "type": "object",
      },
      "UpdateUserProfileVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "data": {
            "additionalProperties": false,
            "description": "返回数据",
            "properties": {
              "userId": {
                "description": "用户ID",
                "examples": [
                  "123e4567-e89b-12d3-a456-426614174000",
                ],
                "format": "uuid",
                "pattern": "^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000)$",
                "type": "string",
              },
            },
            "required": [
              "userId",
            ],
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              true,
            ],
            "type": "boolean",
          },
        },
        "required": [
          "success",
          "data",
        ],
        "type": "object",
      },
      "UserNotFoundVo": {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "additionalProperties": false,
        "properties": {
          "error": {
            "additionalProperties": false,
            "properties": {
              "code": {
                "description": "错误码",
                "examples": [
                  "USER_NOT_FOUND",
                ],
                "type": "string",
              },
              "message": {
                "description": "错误消息",
                "examples": [
                  "指定的用户不存在",
                ],
                "type": "string",
              },
            },
            "required": [
              "code",
              "message",
            ],
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              false,
            ],
            "type": "boolean",
          },
        },
        "required": [
          "success",
          "error",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/v1/users/{userId}/profile": {
      "put": {
        "deprecated": true,
        "description": "更新指定用户的个人档案信息，包括基本信息和偏好设置。支持部分更新，只更新提供的字段。",
        "operationId": "updateUserProfile",
        "parameters": [
          {
            "description": "用户唯一标识符",
            "example": "123e4567-e89b-12d3-a456-426614174000",
            "in": "path",
            "name": "userId",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/UpdateUserProfileDto",
              },
            },
          },
          "description": "用户档案更新数据",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UpdateUserProfileVo",
                },
              },
            },
            "description": "用户档案更新成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserNotFoundVo",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户档案信息",
        "tags": [
          "users",
          "profile",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
    {
      "name": "profile",
    },
  ],
}
`;

exports[`OpenAPIParser > 标签解析功能测试 > 应该解析包含JSON Schema的JSDoc注释 1`] = `
{
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/v1/users/{userId}/profile": {
      "put": {
        "deprecated": true,
        "description": "更新指定用户的个人档案信息，包括基本信息和偏好设置。支持部分更新，只更新提供的字段。",
        "operationId": "updateUserProfile",
        "parameters": [
          {
            "description": "用户唯一标识符",
            "example": "123e4567-e89b-12d3-a456-426614174000",
            "in": "path",
            "name": "userId",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "avatar": {
                    "description": "头像URL",
                    "examples": [
                      "https://cdn.example.com/avatars/user123.jpg",
                    ],
                    "format": "uri",
                    "type": "string",
                  },
                  "displayName": {
                    "description": "用户显示名称",
                    "examples": [
                      "John Doe",
                    ],
                    "maxLength": 50,
                    "minLength": 2,
                    "type": "string",
                  },
                  "email": {
                    "description": "用户邮箱地址",
                    "examples": [
                      "john.doe@example.com",
                    ],
                    "format": "email",
                    "type": "string",
                  },
                },
                "required": [
                  "email",
                ],
                "type": "object",
              },
            },
          },
          "description": "用户档案更新数据",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "返回数据",
                      "properties": {
                        "userId": {
                          "description": "用户ID",
                          "examples": [
                            "123e4567-e89b-12d3-a456-426614174000",
                          ],
                          "format": "uuid",
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                    "success": {
                      "description": "是否成功",
                      "examples": [
                        true,
                      ],
                      "type": "boolean",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "用户档案更新成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "error": {
                      "properties": {
                        "code": {
                          "description": "错误码",
                          "examples": [
                            "USER_NOT_FOUND",
                          ],
                          "type": "string",
                        },
                        "message": {
                          "description": "错误消息",
                          "examples": [
                            "指定的用户不存在",
                          ],
                          "type": "string",
                        },
                      },
                      "type": "object",
                    },
                    "success": {
                      "description": "是否成功",
                      "examples": [
                        false,
                      ],
                      "type": "boolean",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户档案信息",
        "tags": [
          "users",
          "profile",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
    {
      "name": "profile",
    },
  ],
}
`;

exports[`OpenAPIParser > 标签解析功能测试 > 应该解析包含内联schema的JSDoc注释 1`] = `
{
  "components": {
    "schemas": {
      "UpdateUserProfileDto": {
        "properties": {
          "avatar": {
            "description": "头像URL",
            "examples": [
              "https://cdn.example.com/avatars/user123.jpg",
            ],
            "format": "uri",
            "type": "string",
          },
          "displayName": {
            "description": "用户显示名称",
            "examples": [
              "John Doe",
            ],
            "maxLength": 50,
            "minLength": 2,
            "type": "string",
          },
          "email": {
            "description": "用户邮箱地址",
            "examples": [
              "john.doe@example.com",
            ],
            "format": "email",
            "type": "string",
          },
        },
        "required": [
          "email",
        ],
        "type": "object",
      },
      "UpdateUserProfileVo": {
        "properties": {
          "data": {
            "description": "返回数据",
            "properties": {
              "userId": {
                "description": "用户ID",
                "examples": [
                  "123e4567-e89b-12d3-a456-426614174000",
                ],
                "format": "uuid",
                "type": "string",
              },
            },
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              true,
            ],
            "type": "boolean",
          },
        },
        "type": "object",
      },
      "UserNotFoundVo": {
        "properties": {
          "error": {
            "properties": {
              "code": {
                "description": "错误码",
                "examples": [
                  "USER_NOT_FOUND",
                ],
                "type": "string",
              },
              "message": {
                "description": "错误消息",
                "examples": [
                  "指定的用户不存在",
                ],
                "type": "string",
              },
            },
            "type": "object",
          },
          "success": {
            "description": "是否成功",
            "examples": [
              false,
            ],
            "type": "boolean",
          },
        },
        "type": "object",
      },
    },
  },
  "info": {
    "title": "",
    "version": "1.0.0",
  },
  "openapi": "3.1.0",
  "paths": {
    "/api/v1/users/{userId}/profile": {
      "put": {
        "deprecated": true,
        "description": "更新指定用户的个人档案信息，包括基本信息和偏好设置。支持部分更新，只更新提供的字段。",
        "operationId": "updateUserProfile",
        "parameters": [
          {
            "description": "用户唯一标识符",
            "example": "123e4567-e89b-12d3-a456-426614174000",
            "in": "path",
            "name": "userId",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/UpdateUserProfileDto",
              },
            },
          },
          "description": "用户档案更新数据",
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UpdateUserProfileVo",
                },
              },
            },
            "description": "用户档案更新成功",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/UserNotFoundVo",
                },
              },
            },
            "description": "用户不存在",
          },
        },
        "summary": "更新用户档案信息",
        "tags": [
          "users",
          "profile",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "users",
    },
    {
      "name": "profile",
    },
  ],
}
`;
